File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.go
func lengthOfLongestSubstring(s string) int {
    charMap := make(map[byte]int)
	maxLength := float64(0)
	left := float64(0)

	for i := 0; i < len(s); i++ {
		c := s[i]
		if val, ok := charMap[c]; ok {
			left = math.Max(left, float64(val)+1)
		}
		charMap[c] = i
		maxLength = math.Max(maxLength, float64(i)-left+1)
	}

	return int(maxLength)
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_dict = {}
        max_length = 0
        left = 0

        for i, c in enumerate(s):
            if c in char_dict:
                left = max(left, char_dict[c] + 1)
            char_dict[c] = i
            max_length = max(max_length, i - left + 1)
        
        return max_length